ctapipe/image/cleaning.py

```diff
@@ -241,6 +243,56 @@ def apply_time_delta_cleaning(
     return pixels_to_keep
 
 
+def apply_time_average_cleaning(
+    geom, mask, image, arrival_times, picture_thresh, time_limit
+):
+    """
+    Extract all pixels that arrived within a given timeframe 
+    with respect to the time average of the pixels on the main island.
+
+    Parameters
+    ----------
+    geom: `ctapipe.instrument.CameraGeometry`
+        Camera geometry information
+    mask: array, boolean
+        boolean mask of *clean* pixels before time_delta_cleaning
+    image: array
+        pixel values
+    arrival_times: array
+        pixel timing information
+    picture_thresh: float
+        threshold above which time limit is extended twice its value
+    time_limit: int or float
+        arrival time limit w.r.t. the average time of the core pixels
+
+    Returns
+    -------
+
+    A boolean mask of *clean* pixels.  To get a zero-suppressed image and pixel
+    list, use `image[mask], geom.pix_id[mask]`, or to keep the same
+    image size and just set unclean pixels to 0 or similar, use
+    `image[~mask] = 0`
+
+    """
+    pixels_to_remove = []
+    mask = mask.copy()
+    if sum(mask) > 0:
+
+        # use main island (maximum charge) for time average calculation
+        num_islands, island_labels = number_of_islands(geom, mask)
+        mask_main = island_labels == np.argmax([np.sum(image[np.where(island_labels == l)]) for l in range(1,num_islands+1)]) + 1
+        time_ave = np.average(arrival_times[np.where(mask_main)[0]], weights=image[np.where(mask_main)[0]]**2)
+
+        for pixel in np.where(mask)[0]:
```

Comment:
This looks like it can be vectorized. Please check if you can make it work without the for loop just using numpy.

If the problem is the `time_limit`, you can do `limit = np.where(image < (2 * picture_threshold), time_limit, time_limit * 2)`

Reply:
Thank you. I removed for loop using np.where() you suggested.